### ISSUE TITLE
Need help ?

### ISSUE THREAD

**Issue by cuistax** [t-edson/PicPas#37]
Hello all,

I just discovered your project and find it very interesting. 

I would be happy to contribute, I think I could be helpful for:
- Mac target binaries
- French translation (and English)
- Support for new PICs

Besides, I have one question, a friend of mine is using Linux. What's the most recent version he can use ? Any plan to release up to date binaries for that platform ?

Thanks!

**Comment by t-edson** [comment]
Hi.

Thank for test. All collaborations are wellcome. French translation would be excellent for PicPas. Just tell me to add new entries for a new language. And tell me if you have doubts about adding new PICs. It's needed too, to complete the current PIC definitions.

I mostly works only on Windows so I don't compile Linux version for all versions of PicPas.  Just I compile a binary for Ubuntu in some versions. Tha last was for the 0.8.2 version. Anyway I can compile a Ubuntu binary of the currrent version if it's needed.

I don't have a Mac to compile so I cannot test on that. But if you can compile it would be great.

**Comment by cuistax** [comment]
Ok great.

I will look into that. I will get back to you once I get some result.

**Comment by t-edson** [comment]
A binary Ubuntu version is compiled for PicPas 0.8.4: 
![imagen](https://user-images.githubusercontent.com/7736802/38471258-0d2c2264-3b34-11e8-8225-b093814dae3b.png)

**Comment by cuistax** [comment]
Excellent, thanks. I will install that version on my friend's Linux.

I am currently trying to compile the project on my Mac. 
So far the error messages are winning, but the fight is barely started ;)

**Comment by cuistax** [comment]
Hey,

I could get it to work on Mac, where shall I put the binary ?

![picpas_mac_screenshot](https://user-images.githubusercontent.com/12441426/38523948-4803e330-3c4d-11e8-9cbd-9a5d44dca685.jpg)

Please note that I had to update Misutils.pas: it seems that with Lazarus 1.8.2 (latest version), unit FileUtil is not supported anymore, it must be replaced by LazFileUtils.

Thanks,

**Comment by t-edson** [comment]
Excelent. Please put the binary in your GitHub or put it directly here in the PicPas repository and I will commit the request.

**Comment by cuistax** [comment]
There you go. As I stated in the install note, it must yet be considered experimental. 

Many details, such as the app icon, have yet to be sorted.
That said, the app seems to work quite well on my Mac (isn't Lazarus portability great ??)

So, the next release can only be better ;)

**Comment by t-edson** [comment]
Thanks. I will compile the next version using Lazarus 1.8. If you found some details in the code that need to be changed in order to easy the compiling to Mac, just tell me to change. 
Some features like the file explorer, works only on Windows. They need to be extended to Linux and mac too. All other features should work OK in all platforms.

**Comment by cuistax** [comment]
Regarding your last comment, at the moment two things can be done:
- use LazFileUtils instead of FileUtil in Misutils.pas (as mentionned in post nr. 6)
- the project name should be renamed from "project1" to "PicPas" because, as you can see in the Mac screenshot (also in post nr. 6), the name "project1" gets displayed by default in the menu bar.
I can build another Mac binary whenever needed.

**Comment by t-edson** [comment]
OK. I will do the changes.

**Comment by t-edson** [comment]
Ready. New project is now "PicPas", compiled with Lazarus 1.8 and the unit FileUtil is changed. 

This version 0.8.5 use now the library https://github.com/t-edson/ogEditGraf in the version 2.2 (the default), so it needs to be updated to compile the project PicPas.

**Comment by t-edson** [comment]
Many internals changes have been done in the version 0.8.5. I'm closing this version to open a new version.  I'm geenrating the exe for Windows-64 and Ubuntu. Please,generate the binary for Mac in this version. Don't forget to update the library PicUtils.
I hope the 0.8.6 will include finally a compiler for the Baseline family. Let's see what can be done.

**Comment by cuistax** [comment]
Yeah no problem, but I can't seem to compile the project right now.

According to the error log below, there is a missing file:

```
(3104) Compiling FrameArcExplor.pas
/tmp/PicPas/Source/FrameArcExplor.pas(167,4) Fatal: (2013) Cannot open include file "../language/tra_FrameArcExplor.pas"
```

**Comment by t-edson** [comment]
It's my fault. I have created a new translation file. Maybe you can add the french translation to this new file. It's uploaded now.

**Comment by t-edson** [comment]
I have opened the branch 0.8.6 and have the new changes there. Please make the changes on that version.

**Comment by cuistax** [comment]
The Mac binary for branch 0.8.5 is ready : #47

**Comment by t-edson** [comment]
Thanks. I'm glad to say the version 0.8.6 can additionally use the PIC10 compiler so it can compile now for the 12 bits core devices. There are still, some code to modify and some forms to adapt to these new compiler and the simulator is not working by now, but PicPas can now compile a "Hello World" for the PIC10F200.

**Comment by cuistax** [comment]
I would like to compile version 0.8.6 but I get the following errors:

````
FormDebugger.pas(114,7) Error: (5038) identifier idents no member "ExecNCycles"
FormDebugger.pas(115,6) Warning: (5036) Local variable "stopped" does not seem to be initialized
FormDebugger.pas(161,16) Error: (5038) identifier idents no member "DisassemblerAt"
FormDebugger.pas(162,15) Error: (5038) identifier idents no member "ReadPC"
FormDebugger.pas(224,15) Error: (5038) identifier idents no member "ReadPC"
FormDebugger.pas(262,14) Error: (5038) identifier idents no member "DisassemblerAt"
FormDebugger.pas(281,7) Error: (5038) identifier idents no member "Reset"
FormDebugger.pas(295,7) Error: (5038) identifier idents no member "Exec"
FormDebugger.pas(297,7) Error: (5038) identifier idents no member "ExecNCycles"
FormDebugger.pas(298,6) Warning: (5036) Local variable "stopped" does not seem to be initialized
FormDebugger.pas(323,7) Error: (5038) identifier idents no member "WritePC"
FormDebugger.pas(338,7) Error: (5038) identifier idents no member "ExecTo"
FormDebugger.pas(365,7) Error: (5038) identifier idents no member "ExecStep"
FormDebugger.pas(371,7) Error: (5038) identifier idents no member "Exec"
FormDebugger.pas(474) Fatal: (10026) There were 12 errors compiling module, stopping
````

**Comment by t-edson** [comment]
It looks like you have to update PicUtils 1.4. Several changes have been done there.
I just updated PicPas 0.8.6. Now the compiler include a Combo in the IDE to manually switch the compiler to use. The units for devices are separated in /devices10, /devices16, /devices17 (empty) and /devices18 (empty). And the Debugger can now simulate too the 12 bits PICs.
Tell me if you have problems compiling.

**Comment by t-edson** [comment]
There are some issues with the code generator of PIC10 compiler because it use the same scheme of the PIC16 compiler, wich use the FSR register in some cases as an auxiliar register, but in the 12 bits core PICs, FSR is just implemented for some bits. I'm preparing some test for the PIC10 compiler and then the bugs will be detected.

**Comment by cuistax** [comment]
Did you find a solution to that ?

**Comment by t-edson** [comment]
Hi. I'm working on that. There are thousands lines of code to check, identify and change. It's not only for the places where FSR is used; but when the instruction ADDLW or SUBLW are used too. Now the work is in the version 0.8.7 and several changes have made in the internal architecture of the classes to reuse most of the code of the compilers.

**Comment by t-edson** [comment]
Maybe you can collaborate creating the Test code for the Baseline compiler and test them. The files are in /testcode10.
I have an additional problem. I use Proteus to run simulation for the Test code and I have an old version (and cannot update)  and cannot simulate a multi-RAM bank device like the PIC16f57. This limit my possibilities to test the new PIC10 compiler.

**Comment by cuistax** [comment]
Ok I will look into that.

On my side I have several (real) baseline chips that I can use for testing : PIC10F202, PC12F508 and PIC16F506.  

As for the simulation part, you may be interested to know that with Microchip's MPLAB X IDE, you can import external .hex files into a new project and run it on their simulator.
And this is possible with the free version, you don't have to pay anything.

This video from Microchip shows how to import external .hex files : [https://www.youtube.com/watch?v=pEMORwwuyos](https://www.youtube.com/watch?v=pEMORwwuyos).
In this video they only program a chip, but you can simulate it also.

I have already done it with .hex files generated by the SDCC compiler, it works.
Let me know if you need further information on that.

**Comment by t-edson** [comment]
No problem with simulate in MPLAB. PicPas can generate ASM text, totally compatible with MPLAB. That's how I test some codes. 
The problem is the test codes of PicPas is prepared to run in Proteus and have an easy way to run several test per second and detect when some code fails, using sound. Check the Proteus circuit in /testcode10.
Maybe there is a easy way to implement better test codes for MPLAB, but I haven't explored it. 
The idea would be to automate all the tests in a simple Click. Now it takes some time to run all the tests in Proteus and it's a hard work for me. 
I doubt It can be done with MPLAB too, but maybe there is some way to implement similar procedures.

**Comment by t-edson** [comment]
I can test some routines in the same simulator of PicPas and it's the more comfortable way to test. But the PicPas simulator needs to be "validated" before, using some other Simulator. Moreover, there is not an interfaz so powerful like Proteus have. Maybe it's better to improve the PicPas simulator and make all the tests run in just one click.

**Comment by cuistax** [comment]
> Check the Proteus circuit in /testcode10.

Where is it ?

> The idea would be to automate all the tests in a simple Click.

> The problem is the test codes of PicPas is prepared to run in Proteus

> I can test some routines in the same simulator of PicPas

So, if I follow you:
- your tests are currently only usable with Proteus
- you would like to validate both the PicPas simulator and the generated code, but currently only the generated code can be tested
- you would like to use the PicPas simulator to validate the compiled code, and get rid of Proteus
- you would like to get better automation for tests

Is that right ?


So, just to clarify things for me, could you answer the following questions:

1) how exactly would you like to validate the simulator ? what would the process be like ? what is currently missing to achieve that goal (in terms of dev, not in terms of test cases) and how much effort would it require ? 

2) how exactly would you like to validate the compiled code ? what would the process be like ? what is currently missing to achieve that goal and how much effort would it require ? (again, only in terms of dev)

**Comment by t-edson** [comment]
> Where is it ?

I haven't uploaded the folder. It's updated now. 
Here I have prepared a set of test in the same way it exists for the PIC16 compiler in the folder "/TestCode" ( I will change the name to "/testcode16" later) These test are some extense but they are far to be complete.

>    your tests are currently only usable with Proteus

The short answer is YES. They were created for Proteus to test the first versions of PicPas. Maybe these test can be used too in MPLAB but it will take much more time to verify (in the way they are written now). Probably hours. In Proteus they can be tested in minutes.
There is a group of test that can (must) be verified directly in PicPas, like https://github.com/t-edson/PicPas/blob/0.8.7/TestCode/MustPass/Test5_Banks.pas, that use DIRECTIVES to verify the internal variables of the compiler

>    you would like to validate both the PicPas simulator and the generated code, but currently only the generated code can be tested

Debugger/simulator was one of the early objectives of PicPas. It's part of the IDE and like any other tool it must be correctly tested. I have tested by comparing the results when executing code and I must say the PicPas Debugger/simulator is very accurate (I haven't report for bugs). Even it's, in some cases, more exactly than Proteus, but that's another issue. 
The PicPas simulator is very exactly emulating all the instructions of the PIC CPU Core, but it lacks of simulation for peripherals. Timers, Watchdog, UART, ... are not currently simulated. There is some work to do to implement this, and this is compiler developing, this is something different.
 
However I have not a formal way to test the Simulator. I think a formal test must not be so complicated, because PIC Cores are RISC, and it would be focused in check the result for all the instructions (in results and in cycles used)

>    you would like to use the PicPas simulator to validate the compiled code, and get rid of Proteus

Yes.

>    you would like to get better automation for tests

Yes.

**Comment by t-edson** [comment]
>  how exactly would you like to validate the simulator ? what would the process be like ? what is currently missing to achieve that goal (in terms of dev, not in terms of test cases) and how much effort would it require ?

As far as possible for a free project. Including all the peripherals implemented in the simulator (Currently none).
There is needed to create some test for that. I have not clear what would be the best way for test (I have mentioned some clue in my last comment). Fiinally, all of this is just development (and time) someone needs to do, but I have not idea how much effort or time it would require.

>  how exactly would you like to validate the compiled code ? what would the process be like ? what is currently missing to achieve that goal and how much effort would it require ? (again, only in terms of dev)

As far as possible for a free project. The process here is: Prepare (and execute) the test code. This test code must be detailed in test all the use cases of the compiler and here there are some problem:

* The obviouss: Create extense and detailed code is a heavy work nobody wants to do
* Extense code cannot fit in the small memory of the devices (specially the PIC10 series). So code must be splitted in different programs that can be tested individually, making the process more heavy and long.

That's the reason the testing must be automated. Modifyng some cirtical routines in the code generator can affect many points of the compiler. Changing just one line of code can force to run all the batteries of tests, and if takes 20 or 30 minutes, it's not productive.

**Comment by cuistax** [comment]
Well, to start somewhere, I believe the first necessity would be to have a test framework for the simulator.

At the very least, all you need is a function which takes in input:
- an initial memory content (code+data)
- an expected memory content (data only) after the input program has run

The function takes the input code+data, runs it into the simulator (the code must terminate itself), and once simulation is done, compares the obtained memory content with the expected one, and returns a validation status.

You can always refine it later to take the number of cycles into account and that sort of things, but that would get you started.

How much effort would that take ?

**Comment by t-edson** [comment]
Basically the test for Simulator would be one more program to simulate. The idea would be to use all the instructions and check the result, the state of the flags, and the cycles used. It' would be interesting doing a consistent program using all the instruction of one family. The test should include invalid Opcodes too, to verify the response.
I think this will require to add some features to the simulator like components (similar to a led)s that can be conected to a pin of the PIC and generate some control action (like  Stop Simulation, Reset device, Show a Message box, write a text in the message window, generate a Sound, ... ).
![imagen](https://user-images.githubusercontent.com/7736802/40869855-18fa7fda-65e8-11e8-94aa-2f4cf1dcc79b.png)
This is a lot of work (some kind of visual programming) but I think this is the best way to proceed.
It's projected to include some LCD and Serial Terminal to the simulator but this devices are jus part of the hardware simualtion and what we need is a better control of the simulation.
The detection of errors will: 
* Stop the simulation
* Locate the point of the simulation where it fails (and better the point of the source code).

Doing all this hard work it's only the first phase of automation testing (assuming we have migrated all the test code) and will let us prescind of Proteus. 
The next phase would be the real automation that is the possibility to run several programs with just one click or command. This would involucrate some kind of script or code that simulate sequentially many files and stop when some fails, showing the file and the point where the test failed.

**Comment by t-edson** [comment]
Hi. I've added several functions to the simulator of PicPas. Now it's possible to change the state of pins of the microcontroller and connect the output to simple components: 
![imagen](https://user-images.githubusercontent.com/7736802/41204104-def1d756-6ca5-11e8-83cf-c4de3fa74a88.png)
Several changes has been done in this version. I'm going to close the 0.8.7. Please generate the executable for Mac. You need to update https://github.com/t-edson/PicUtils and https://github.com/t-edson/ogEditGraf

**Comment by cuistax** [comment]
That looks really pretty !
I will check it out.

And yes, I will provide an updated Mac binary soon...

**Comment by alonsobsd** [comment]
Hi everyone

I've added picpas to FreeBSD ports tree. 

https://www.freshports.org/editors/picpas/

Maybe, you could look at the following patch files that I have added

https://svnweb.freebsd.org/ports/head/editors/picpas/files/patch-Source_Globales.pas?revision=476809&view=markup (Add basic support for config files into homedir)
https://svnweb.freebsd.org/ports/head/editors/picpas/files/patch-language_tra_FrameArcExplor.pas?view=markup (Fix a compiling problem)

Tito saludos desde Perú

Greetings

**Comment by t-edson** [comment]
Hi @jacardenasm 

It's good to know that PicPas is available in FreeBSD now. 

Just one detail; PicPas is now growing to include the 10F and the Enhanced PIC16F families, so a best title could be "Multi-platform Pascal compiler for Microchip 8 bits PIC microcontrollers". 

Saludos.
